lib/routemaster/api_client.rb

```diff
@@ -25,16 +25,23 @@
 
 module Routemaster
   class APIClient
+
+    # Memoize the root resources at Class level so that we don't hit the cache
+    # all the time to fetch the root resource before doing anything else.
+    @@root_resources = {}
+
     def initialize(middlewares: [],
                    listener: nil,
                    response_class: nil,
                    metrics_client: nil,
                    source_peer: nil)
-      @listener = listener
-      @middlewares = middlewares
+
+      @listener       = listener
+      @middlewares    = middlewares
       @response_class = response_class
       @metrics_client = metrics_client
-      @source_peer = source_peer
+      @source_peer    = source_peer
+      @lock           = Mutex.new
```

Comment:
Sorry, I didn't realize that in the second snipped you should not use `MonitorMixin`. My bad.